connector/src/main/java/org/geysermc/connector/network/translators/java/JavaKeepAliveTranslator.java

```diff
@@ -25,24 +25,26 @@
 
 package org.geysermc.connector.network.translators.java;
 
+import com.github.steveice10.mc.protocol.packet.ingame.client.ClientKeepAlivePacket;
 import com.github.steveice10.mc.protocol.packet.ingame.server.ServerKeepAlivePacket;
 import com.nukkitx.protocol.bedrock.packet.NetworkStackLatencyPacket;
 import org.geysermc.connector.network.session.GeyserSession;
 import org.geysermc.connector.network.translators.PacketTranslator;
 import org.geysermc.connector.network.translators.Translator;
 
+import java.util.concurrent.TimeUnit;
+
 /**
- * Used to forward the keep alive packet to the client in order to get back a reliable ping.
+ * Used to send the keep alive packet back after delaying it by the Bedrock client's latency
  */
 @Translator(packet = ServerKeepAlivePacket.class)
 public class JavaKeepAliveTranslator extends PacketTranslator<ServerKeepAlivePacket> {
 
     @Override
     public void translate(ServerKeepAlivePacket packet, GeyserSession session) {
-        session.setLastKeepAliveTimestamp(packet.getPingId());
-        NetworkStackLatencyPacket latencyPacket = new NetworkStackLatencyPacket();
-        latencyPacket.setFromServer(true);
-        latencyPacket.setTimestamp(packet.getPingId());
-        session.sendUpstreamPacket(latencyPacket);
+        // Delay sending the keep alive response packet by the Bedrock client's latency for more accurate ping
+        session.getConnector().getGeneralThreadPool().schedule(() ->
+                session.sendDownstreamPacket(new ClientKeepAlivePacket(packet.getPingId())),
+                session.getUpstream().getSession().getLatency(), TimeUnit.MILLISECONDS);
```

Comment:
Shouldn't this be capped/checked in some way? I don't know about in game but in server list I've seen a latency of something like -2,147,483,647 or something quite often

Reply:
I think that's its own bug. I've heard Redned calling it such, at least.

Reply:
But still, it doesn't make sense to delay an arbitrary amount of time if the latency is miscalculated (there can be huge spikes)... I think it should be capped at around 1 or 2 seconds